## Uncovering underlying physical principles and driving forces of cell differentiation & reprogramming using single-cell transcriptomics

![image](https://github.com/Zhu-1998/celldevelopment/blob/main/Graph.jpg)

Advances in single-cell technology enable transcriptome data at unprecedented scales. However, identifying the driving force of cell function from these data remains challenging. We address this by learning cell state vector fields of cell differentiation/reprogramming from single-cell RNA velocity to quantify the global driving forces as landscape and flux. Our findings show that stem/progenitor cells necessitate greater energy dissipation for rapid cell cycles and self-renewal, maintaining pluripotency. We predict optimal developmental pathways and uncover the nucleation mechanism of cell state phase transitions with transition states as nucleation sites and pioneer genes as nucleation seeds. Loop flux allows quantification of each cycle flux's contribution to cell state transition dynamics, offering optimized bio-functions. We infer cell-cell interactions and cell-type-specific gene regulatory networks from single-cell data, predicting the effects of genetic perturbations on cell fate decisions. Essentially, our methodology utilizes single-cell high-throughput experiments for validating the Waddington landscape and flux theory, with associated quantifications. This provides a novel framework for uncovering underlying physical principles of cellular differentiation/reprogramming and broader biological processes.
